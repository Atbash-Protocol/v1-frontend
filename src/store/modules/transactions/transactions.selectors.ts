import { IReduxState } from 'store/slices/state.interface';

import { TransactionType } from './transactions.type';

export const selectPendingTx = (state: IReduxState, type: TransactionType): boolean => {
    const tx = state.transactions.find(t => t.type === type);

    return !tx ? false : true;
};

export const selectTransactionPending = (state: IReduxState) => state.transactions.length > 0;
