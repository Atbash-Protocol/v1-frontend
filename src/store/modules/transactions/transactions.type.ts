export const TransactionTypeEnum = {
    APPROVE_CONTRACT: 'APPROVE_CONTRACT',
    STAKE_PENDING: 'STAKE_PENDING',
    UNSTAKE_PENDING: 'UNSTAKE_PENDING',
    BASH_APPROVAL: 'BASH_APPROVAL',
    SBASH_APPROVAL: 'SBASH_APPROVAL',
    BONDING: 'BONDING',
    REDEEMING: 'REDEEMING',
    REDEEMING_STAKING: 'REDEEMING_STAKING',
} as const;

export type TransactionType = typeof TransactionTypeEnum[keyof typeof TransactionTypeEnum];

export interface Transaction {
    hash: string;
    type: TransactionType;
    status: 'PENDING' | 'DONE';
}
