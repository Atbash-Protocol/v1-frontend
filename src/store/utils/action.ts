import { AnyAction, Action } from '@reduxjs/toolkit';

interface RejectedAction extends Action {
    error: Error;
}

export const isActionRejected = (action: AnyAction): action is RejectedAction => action.type.endsWith('rejected');
