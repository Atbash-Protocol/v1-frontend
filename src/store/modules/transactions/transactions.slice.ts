import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { Transaction } from './transactions.type';

const initialState = new Array<Transaction>();

const pendingTxnsSlice = createSlice({
    name: 'pendingTransactions',
    initialState,
    reducers: {
        addPendingTransaction(state, action: PayloadAction<Pick<Transaction, 'hash' | 'type'>>) {
            state.push({ ...action.payload, status: 'PENDING' });
        },
        clearPendingTransaction(state, action: PayloadAction<string>) {
            const target = state.find(x => x.hash === action.payload);
            if (target) {
                target.status = 'DONE';
            }
        },
    },
});

export const { addPendingTransaction, clearPendingTransaction } = pendingTxnsSlice.actions;

export default pendingTxnsSlice.reducer;
