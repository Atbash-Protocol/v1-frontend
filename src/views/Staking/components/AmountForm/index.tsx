import { useCallback, useState, MouseEvent, ChangeEvent } from 'react';

import { Box, Button, CircularProgress, Grid, InputAdornment, OutlinedInput, Typography } from '@mui/material';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';

import { theme } from 'constants/theme';
import { addNotification } from 'store/modules/messages/messages.slice';
import { selectPendingTx } from 'store/modules/transactions/transactions.selectors';
import { TransactionType } from 'store/modules/transactions/transactions.type';
import { IReduxState } from 'store/slices/state.interface';

interface AmountFormProps {
    initialValue: number;
    maxValue: number;
    transactionType: TransactionType;
    approvesNeeded: boolean;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    onApprove: any;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    onAction: any;
    approveLabel: string;
    actionLabel: string;
    isLoading?: boolean;
}

const AmountForm = (props: AmountFormProps) => {
    const { t } = useTranslation();
    const dispatch = useDispatch();
    const { initialValue, maxValue, transactionType, approvesNeeded, onApprove, onAction, approveLabel, actionLabel, isLoading } = props;

    const [value, setValue] = useState<string>('');
    const selectPendingTransaction = useSelector<IReduxState, boolean>(state => selectPendingTx(state, transactionType));

    const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
        e.preventDefault();
        setValue(e.target.value);
    };

    const handleClickMaxValue = () => {
        setValue(maxValue.toString());
    };

    const handleActionClick = useCallback(
        (e: MouseEvent) => {
            e.preventDefault();

            if (selectPendingTransaction) return;

            if (approvesNeeded) {
                return onApprove(transactionType);
            }

            if (value === '0') {
                return dispatch(addNotification({ severity: 'warning', description: 'Please provide an amount' }));
            }

            return onAction(value);
        },
        [onApprove, onAction, value, transactionType],
    );

    return (
        <Grid container>
            <Grid item xs={10}>
                <OutlinedInput
                    sx={{
                        color: theme.palette.primary.main,
                        border: '1px solid',
                        borderColor: theme.palette.primary.main,
                        borderRadius: 0,
                        outlineColor: theme.palette.primary.main,
                        borderRight: 'none',
                        width: '100%',
                    }}
                    type="number"
                    placeholder={initialValue.toString()}
                    value={value}
                    onChange={handleChange}
                    inputProps={{ inputMode: 'numeric', pattern: '^[0-9]*[.,]?[0-9]*$' }}
                    endAdornment={
                        <InputAdornment position="end">
                            <Box sx={{ color: theme.palette.primary.main, textTransform: 'uppercase', cursor: 'pointer' }} onClick={handleClickMaxValue}>
                                <Typography>
                                    <>{t('Max')}</>
                                </Typography>
                            </Box>
                        </InputAdornment>
                    }
                />
            </Grid>
            <Grid item xs={2} p={0}>
                <Button
                    sx={{
                        padding: 0,
                        color: theme.palette.primary.main,
                    }}
                    disabled={isLoading}
                    onClick={handleActionClick}
                >
                    {isLoading && <CircularProgress color="secondary" />}

                    {!isLoading && (
                        <Typography
                            variant="body1"
                            sx={{
                                [theme.breakpoints.down('sm')]: {
                                    maxWidth: '9ch',
                                    padding: 0,
                                    overflow: 'hidden',
                                    fontSize: '.5rem',
                                    whiteSpace: 'nowrap',
                                },
                            }}
                        >
                            {' '}
                            {approvesNeeded ? approveLabel : actionLabel}
                        </Typography>
                    )}
                </Button>
            </Grid>
        </Grid>
    );
};

export default AmountForm;
