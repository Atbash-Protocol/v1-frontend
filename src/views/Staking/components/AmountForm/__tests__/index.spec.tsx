import { configureStore } from '@reduxjs/toolkit';

import { TransactionTypeEnum } from 'store/modules/transactions/transactions.type';
import { renderComponent } from 'tests/utils';

import AmountForm from '..';

describe('#AmountForm', () => {
    it('renders', () => {
        const props = {
            initialValue: 0,
            maxValue: 10,
            transactionType: TransactionTypeEnum.APPROVE_CONTRACT,
            approvesNeeded: true,
            onApprove: jest.fn(),
            onAction: jest.fn(),
            approveLabel: 'approveLabel',
            actionLabel: 'actionLabel',
            isLoading: true,
        };
        const comp = renderComponent({
            component: <AmountForm {...props} />,
            store: configureStore({
                reducer: {},
            }),
            context: {},
        });

        expect(comp).toMatchSnapshot();
    });
});
