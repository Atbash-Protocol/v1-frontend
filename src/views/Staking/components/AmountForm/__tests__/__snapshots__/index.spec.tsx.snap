// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#AmountForm renders 1`] = `
<Context.Provider
  value={Object {}}
>
  <Provider
    store={
      Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <AmountForm
      actionLabel="actionLabel"
      approveLabel="approveLabel"
      approvesNeeded={true}
      initialValue={0}
      isLoading={true}
      maxValue={10}
      onAction={[MockFunction]}
      onApprove={[MockFunction]}
      transactionType="APPROVE_CONTRACT"
    />
  </Provider>
</Context.Provider>
`;
